runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -0,0 +1,279 @@
+
+            if (bufferSize == DEFAULT_BUFFER_SIZE) {
+                corfuStore.getRuntime().getTableRegistry().getStreamingManager().subscribeAcrossNamespaces(clientListener,
+                        namespace, streamTag, tablesOfInterest, timestamp.getSequence());
+            } else {
+                corfuStore.getRuntime().getTableRegistry().getStreamingManager().subscribeAcrossNamespaces(clientListener,
+                        namespace, streamTag, tablesOfInterest, timestamp.getSequence(), bufferSize);
+            }
+        }
+
+        /**
+         * Returns a valid timestamp at which the client stream listener can be subscribed.  A valid timestamp is one
+         * which is outside the bounds of an ongoing full-sync.   Client callbacks are also invoked to ensure that
+         * updates upto this timestamp have been read and processed.
+         */
+        private CorfuStoreMetadata.Timestamp getValidSubscriptionTimestamp() {
+
+            // Number of milliseconds to wait before each retry
+            int backoffRetryDurationMs = 5;
+
+            // Max duration at which the wait time must be capped (2 minutes)
+            int maxDurationThresholdMins = 2;
+
+            // Retry until a timeout of max_duration_threshold_mins
+            try {
+                return IRetry.build(ExponentialBackoffRetry.class, () -> {
+                    try {
+                        // Check if snapshot sync is currently in progress.  If in progress, wait for it to finish
+                        CorfuStoreMetadata.Timestamp latestTimestamp = getTimestamp();
+                        if (checkSnapshotSyncOngoing(latestTimestamp)) {
+                            waitSnapshotSyncCompletion(latestTimestamp);
+                        }
+
+                        // Invoke the client callback for multi table reads.  This will read all tables the client is
+                        // interested in and return a timestamp.  Verify if snapshot sync was in progress at that
+                        // timestamp.  If in progress, wait for it to complete and retry the whole workflow.
+                        CorfuStoreMetadata.Timestamp multiTableReadTimestamp = clientListener.performMultiTableReads();
+                        if (checkSnapshotSyncOngoing(multiTableReadTimestamp)) {
+                            waitSnapshotSyncCompletion(multiTableReadTimestamp);
+                            log.info("Snapshot sync was going on during multi table read.  Rerun the checks and " +
+                                "re-trigger multiTable read");
+                            throw new RetryNeededException();
+                        }
+
+                        // Snapshot sync was not in progress at the client's read timestamp.  Invoke the client callback
+                        // to set the baseline at this timestamp.
+                        clientListener.mergeTableOnSubscription(multiTableReadTimestamp);
```

Comment:
naming here i a bit off a technically subscription has not happened yet, isn't mergeTables() enough? or mergeReplicatedTables()

Reply:
but do we have an explicit callback for that? no, right? isn't the callback in those cases onSnapshotSyncCompleted?